system/lib/compiler-rt/lib/ubsan_minimal/sanitizer_atomic.h

```diff
@@ -0,0 +1,236 @@
+/**
+ * This file is based on sanitizer_atomic.h, sanitizer_atomic_clang.h, and
+ * sanitizer_atomic_clang_other.h from compiler-rt.
+ * Last changed revision: 351636.
+ * Last changed date: 2019-01-19.
+ *
+ * The files are combined into one to avoid depending on many unneeded headers
+ * in compiler-rt/lib/sanitizer_common.
+ */
+
+//===-- sanitizer_atomic.h --------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file is a part of ThreadSanitizer/AddressSanitizer runtime.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SANITIZER_ATOMIC_H
+#define SANITIZER_ATOMIC_H
+
+#include <cinttypes>
+
+#ifndef INLINE
+#define INLINE inline
+#endif
+#define ALIGNED(x) __attribute__((aligned(x)))
+#define DCHECK(a)
+
+namespace __sanitizer {
+
+enum memory_order {
+  memory_order_relaxed = 1 << 0,
+  memory_order_consume = 1 << 1,
+  memory_order_acquire = 1 << 2,
+  memory_order_release = 1 << 3,
+  memory_order_acq_rel = 1 << 4,
+  memory_order_seq_cst = 1 << 5
+};
+
+struct atomic_uint8_t {
+  typedef std::uint8_t Type;
+  volatile Type val_dont_use;
+};
+
+struct atomic_uint16_t {
+  typedef std::uint16_t Type;
+  volatile Type val_dont_use;
+};
+
+struct atomic_sint32_t {
+  typedef std::int32_t Type;
+  volatile Type val_dont_use;
+};
+
+struct atomic_uint32_t {
+  typedef std::uint32_t Type;
+  volatile Type val_dont_use;
+};
+
+struct atomic_uint64_t {
+  typedef std::uint64_t Type;
+  // On 32-bit platforms u64 is not necessary aligned on 8 bytes.
+  volatile ALIGNED(8) Type val_dont_use;
+};
+
+struct atomic_uintptr_t {
+  typedef std::uintptr_t Type;
+  volatile Type val_dont_use;
+};
+
+typedef std::uintptr_t uptr;
+
+INLINE void proc_yield(int cnt) {
+  __asm__ __volatile__("" ::: "memory");
+}
+
+template<typename T>
+INLINE typename T::Type atomic_load(
+    const volatile T *a, memory_order mo) {
+  DCHECK(mo & (memory_order_relaxed | memory_order_consume
+      | memory_order_acquire | memory_order_seq_cst));
+  DCHECK(!((uptr)a % sizeof(*a)));
+  typename T::Type v;
+
+  if (sizeof(*a) < 8 || sizeof(void*) == 8) {
+    // Assume that aligned loads are atomic.
+    if (mo == memory_order_relaxed) {
+      v = a->val_dont_use;
+    } else if (mo == memory_order_consume) {
+      // Assume that processor respects data dependencies
+      // (and that compiler won't break them).
+      __asm__ __volatile__("" ::: "memory");
```

Comment:
I'm surprised this actually compiles. Is this template ever instantiated?

Reply:
It turns out this compiles because it does not lower to any instructions. TIL.

Reply:
If this file is copy of an upstream file I would leave it unmodified if possible.  Makes it easier to update compiler-rt in the future.